checks/docker/leaked_secrets.rego

```diff
@@ -0,0 +1,192 @@
+# METADATA
+# title: Secrets passed via `build-args` or envs or copied secret files
+# description: Passing secrets via `build-args` or envs or copying secret files can leak them out
+# schemas:
+# - input: schema["dockerfile"]
+# related_resources:
+# - https://docs.docker.com/build/building/secrets/
+# custom:
+#   id: DS031
+#   avd_id: AVD-DS-0031
+#   severity: CRITICAL
+#   short_code: do-not-pass-secrets
+#   recommended_action: Use secret mount if secrets are needed during image build. Use volume mount if secret files are needed during container runtime.
+#   input:
+#     selector:
+#     - type: dockerfile
+package builtin.dockerfile.DS031
+
+import rego.v1
+
+import data.lib.docker
+
+final_stage := last(input.Stages)
+
+# check if env or arg contains secret env
+deny contains res if {
+	some instruction in final_stage.Commands
+	is_arg_or_env(instruction.Cmd)
+	[name, _] := retrive_name_and_default(instruction)
+	is_secret_env(name)
+	res := result.new(
+		sprintf("Possible exposure of secret env %q in %s", [name, upper(instruction.Cmd)]),
+		instruction,
+	)
+}
+
+# check if env or arg contains secret file env
+deny contains res if {
+	some instruction in final_stage.Commands
+	is_arg_or_env(instruction.Cmd)
+	[name, path] := retrive_name_and_default(instruction)
+	path != ""
+	name in secret_file_envs
+	is_secret_file_copied(path)
+	res := result.new(
+		sprintf("Possible exposure of the copied secret env file %q in %s", [name, upper(instruction.Cmd)]),
+		instruction,
+	)
+}
+
+# check if a secret file is copied
+deny contains res if {
+	some instruction in final_stage.Commands
+	instruction.Cmd == "copy"
+	count(instruction.Value) == 2
+	env := trim_prefix(instruction.Value[1], "$")
+	env in secret_file_envs
+	res := result.new(
+		sprintf("Possible exposure of secret file %q in COPY", [env]),
+		instruction,
+	)
+}
+
+check_args := true
+
+# TODO: Should arguments be checked?
+is_arg_or_env(cmd) if {
+	check_args
+	cmd == "arg"
+}
```

Comment:
what's a case that you are thinking to check the args for?

Reply:
Made it enabled by default https://github.com/aquasecurity/trivy-checks/pull/265/commits/9fc294b5d2bbd232097b12c9d92e3bbf23d2eeef